### PR TITLE
Physical server provisioning via StateMachine

### PR DESCRIPTION
**Pull request by miha-plesko**
With this commit we make sure that AE engine calculates correct    entrypoint when PhysicalServer provisioning request is created,     namely:
    
```
PhysicalInfrastructure/PhysicalServer/Lifecycle/Provisioning
```
    
Furthermore, we enhance PhysicalServer AE service model with `ems_type`     method that is used for execution routing inside StateMachine     (i.e. via messages).

Depends on https://github.com/ManageIQ/manageiq/pull/18573
Related to https://github.com/ManageIQ/manageiq-content/pull/516

@miq-bot assign @gmcculloug 
@miq-bot add_label enhancement

### CODE REVIEW

**Comment by miha-plesko** [comment]
Retrying add_label and assign:

@miq-bot add_label enhancement
@miq-bot assign @gmcculloug

File: lib/miq_automation_engine/engine/miq_ae_engine/miq_ae_builtin_method.rb [hunk 1]
```diff
@@ -182,11 +185,15 @@ def self.detect_category(obj_name, prov_obj)
       when "orchestration_stack", "orchestration_stack_retire_task"
         CLOUD
       when "miq_request", "miq_provision", "vm_migrate_task", "vm_retire_task"
+        return PHYSICAL_INFRA if (prov_obj&.options || {})[:request_type] == 'provision_physical_server'
```
Comment (gmcculloug) [inline lib/miq_automation_engine/engine/miq_ae_engine/miq_ae_builtin_method.rb]: Instead of checking the options hash, could we verify they sub-class of object?  Same for the `miq_provision_task` case below.  Event though these are different objects, one request and one task, maybe we could pull this logic into a `physical_server?` method here to DRY up the logic.

File: lib/miq_automation_engine/engine/miq_ae_engine/miq_ae_builtin_method.rb [hunk 2]
```diff
@@ -10,10 +10,12 @@ class MiqAeBuiltinMethod
       vm_retire_task
       service_retire_task
       orchestration_stack_retire_task
+      miq_provision_task
```
[the remark this answers is not in the document]
Comment (miha-plesko) [reply lib/miq_automation_engine/engine/miq_ae_engine/miq_ae_builtin_method.rb]: As discussed on gitter I've now improved this part: `PhysicalServerProvisionTask` now resolves to `physical_server_provision_task` so that it's more obvious what we're provisioning. This also simplifies the IF statements below because task type is now unique.

File: lib/miq_automation_engine/engine/miq_ae_engine/miq_ae_builtin_method.rb (hunk 1 above)
Comment (miha-plesko) [reply lib/miq_automation_engine/engine/miq_ae_engine/miq_ae_builtin_method.rb]: Good point - turns out it's now even more simple because we've separated task name. For MiqRequest I've done it with case statement, please let me know if you find it readable enough. If not, I'll use auxilary function as you suggest.

File: lib/miq_automation_engine/service_models/miq_ae_service_physical_server.rb [hunk 3]
```diff
@@ -5,5 +5,9 @@ class MiqAeServicePhysicalServer < MiqAeServiceModelBase
     expose :turn_off_loc_led
     expose :power_on
     expose :power_off
+
+    def ems_type
+      ar_method { @object.ext_management_system.class.ems_type }
```
[the remark this answers is not in the document]
Comment (miha-plesko) [reply lib/miq_automation_engine/service_models/miq_ae_service_physical_server.rb]: Good point, done. I like your suggestion because we're now able to easier access that value even ouside of automation.

**Comment by miha-plesko** [comment]
FYI Travis is failing because we need to merge core PR first https://github.com/ManageIQ/manageiq/pull/18573. @agrare gave us the thumbsup so @gmcculloug it's all yours :)

File: lib/miq_automation_engine/service_models/miq_ae_service_physical_server_provision_task.rb [hunk 4]
```diff
@@ -0,0 +1,4 @@
+module MiqAeMethodService
+  class MiqAeServicePhysicalServerProvisionTask < MiqAeServiceMiqProvisionTask
+  end
+end
```
Comment (lfu) [inline lib/miq_automation_engine/service_models/miq_ae_service_physical_server_provision_task.rb]: Is this file necessary? @gmcculloug

Comment (gmcculloug) [reply lib/miq_automation_engine/service_models/miq_ae_service_physical_server_provision_task.rb]: No, automate will dynamically create service models and expose columns, virtual_columns and associations from the active_record model.  You only need to create these models now if you modify or expose additional functionality.

Comment (miha-plesko) [reply lib/miq_automation_engine/service_models/miq_ae_service_physical_server_provision_task.rb]: Which I do - I extend `MiqAeServiceMiqProvisionTask` which have a bunch of custom methods like state_machine_status and set_option() that we need.

Comment (gmcculloug) [reply lib/miq_automation_engine/service_models/miq_ae_service_physical_server_provision_task.rb]: The point being this model is empty so those methods are either being exposed automatically or pulled in from sub-classes.  If there is nothing defined here the empty class is not needed.

Comment (miha-plesko) [reply lib/miq_automation_engine/service_models/miq_ae_service_physical_server_provision_task.rb]: I think it didn't work until I added the definition here because [this method](https://github.com/ManageIQ/manageiq-automation_engine/blob/master/lib/miq_automation_engine/service_models/miq_ae_service_miq_provision_task.rb#L3) was not defined  but let me try again tomorrow, probably I messed something up. Anyhow, thanks for pointing up, I had no idea service models are able to automatically inherit between themselves based on actual model relations.

Comment (gmcculloug) [reply lib/miq_automation_engine/service_models/miq_ae_service_physical_server_provision_task.rb]: If you are interested in learning more check out https://github.com/ManageIQ/manageiq/pull/14465 and https://github.com/ManageIQ/manageiq-automation_engine/pull/308

Comment (miha-plesko) [reply lib/miq_automation_engine/service_models/miq_ae_service_physical_server_provision_task.rb]: Surprisingly, you were right again 👍 Works like a charm without the file. Thanks for pointers to PRs

File: lib/miq_automation_engine/engine/miq_ae_engine/miq_ae_builtin_method.rb [hunk 5]
```diff
@@ -181,12 +184,23 @@ def self.detect_category(obj_name, prov_obj)
       case obj_name
       when "orchestration_stack", "orchestration_stack_retire_task"
         CLOUD
-      when "miq_request", "miq_provision", "vm_migrate_task", "vm_retire_task"
+      when "miq_request"
+        case prov_obj
+        when nil
+          UNKNOWN
```
Comment (lfu) [inline lib/miq_automation_engine/engine/miq_ae_engine/miq_ae_builtin_method.rb]: With ```obj_name = 'miq_request'```, ```prov_obj = nil```,  seems old logic would return ```nil```. Is it right?

Comment (miha-plesko) [reply lib/miq_automation_engine/engine/miq_ae_engine/miq_ae_builtin_method.rb]: Yes, but I think it's a bug and should be unknown. But you're right, I'll flip back to nil just to be safe.

**Comment by gmcculloug** [comment]
Close/re-open to resolve failing tests.

**Comment by miha-plesko** [comment]
I think it'll fail until core PR is merged where we define some factories
